modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Emitter.scala

```diff
@@ -38,20 +26,288 @@ trait UUIDProvider[F[_]] {
  *
  */
 trait Emitter[F[_]] {
-  import Emitter._
 
   /**
-   * Emits the event payload
+   * Requests the emitter to send the event to the snowplow collector
    *
+   * Depending on the emitter configuration, this could either send immediately or buffer the event
+   * in a queue to be sent at a later time
    * @param event Fully assembled event
    */
-  def send(event: EmitterPayload): F[Unit]
+  def send(event: Payload): F[Unit]
+
+  /**
+   * Flush pending events to the collector early, before the buffer is full.
+   */
+  def flushBuffer(): F[Unit]
 
 }
 
 object Emitter {
 
-  /** Low-level representation of event */
-  type EmitterPayload = Map[String, String]
+  /** User-provided callback */
+  type Callback[F[_]] = (EndpointParams, Request, Result) => F[Unit]
+
+  /** Emitter buffering config */
+  sealed trait BufferConfig extends Product with Serializable
+
+  object BufferConfig {
+
+    /** Configures the emitter to send events immediately to the collector, without buffering for larger batches */
+    case object NoBuffering extends BufferConfig
+
+    /** Configures the emitter to buffer events and send batched payloads comprising a fixed number of events
+     *  @param size The number of events after which the emitter sends the buffered events to the collector
+     */
+    case class EventsCardinality(size: Int) extends BufferConfig
+
+    /** Configures the emitter to buffer events and send batched payloads of a minimum size in bytes
+     *  @param bytes The target minimum size in bytes of a payload of batched events
+     */
+    case class PayloadSize(bytes: Int) extends BufferConfig
+
+    /** The default [[BufferConfig]]: Events are buffered until a payload size of 40 kb */
+    val Default: BufferConfig = PayloadSize(40000)
+  }
+
+  /** Payload (either GET or POST) ready to be send to collector */
+  sealed trait Request extends Product with Serializable {
+
+    /** Attempt to send */
+    def attempt: Int
```

Comment:
I think we need to clarify that it's _next attempt_, not _amount of times we've tried to submit it_, i.e. it's 1-based, not 0-based. Have no idea how to phrase it myself.